Visualization-First Display functionality

dannon opened the pull request:
Builds on #20154 (which needs tests fixed still, I'll rebase this as needed)

This allows admins to configure preferred visualizations for specific datatypes.  A good example use-case here is h5.  Right now the 'Preview' view of this is just binary file contents.  This allows us to say "h5web" is how we should display this.  Here's a quick example:


https://github.com/user-attachments/assets/8dfc02e8-f955-47e9-b07d-d9840fef0bb6





## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

dannon committed:
Sketch out initial dataset view component

dannon committed:
Add view to itemurls type, incremental refactoring

dannon committed:
fix preview iframe sizing

dannon committed:
Refactor dataset preview tab styling and remove unnecessary CSS rules

dannon committed:
Refine dataset component attribute display

dannon committed:
Fix linting post-rebase

dannon committed:
Adjust routing for dataset view

dannon committed:
streamline tab management and improve error handling

dannon committed:
Consolidate dataset view routes

dannon committed:
Typing, cleanup, formatting, etc. in DatasetView

dannon committed:
Add testing for datasetview component

dannon committed:
Fix typing for datasetview test?

dannon committed:
Dump ts for testing datasetview; too much aggravation to work around for little benefit to actually *testing*

dannon committed:
Cleanup in DatasetActions

dannon committed:
Handle rerun; cleanup in contentOptions

dannon committed:
Preferred viz rough sketch

dannon committed:
Fix showError and onError functions in DatasetActions component

dannon committed:
Adjust datatypes preferred viz API /datatypes/{type}/visualizations

dannon committed:
Remove DatatypeVisualizations component and its route from admin routes

dannon committed:
rebase into endpoint adjustment

dannon committed:
Adds a new `/api/datatypes/{datatype}` endpoint

provides comprehensive information about a specific datatype in a single
response, consolidating information previously spread across multiple
endpoints, and adds a bit new.

dannon committed:
Remove update and delete operations from datatype visualizations store -- bailing on admin control of this for now.

dannon committed:
Update client-api-schema

dannon committed:
Formatting.

mvdbeek commented on lib/galaxy/config/sample/datatypes_conf.xml.sample:
```diff
@@ -1035,6 +1035,23 @@
+      - visualization: The visualization plugin to use
+      - default_params: Optional JSON-formatted parameters to pass to the visualization
+    -->
+    <visualization_mapping datatype="h5" visualization="vitessce" />
```
Didn't you want to use h5web for h5 ?

dannon replied:
Yes, and I think I want to restructure this to shift the mappings up into the individual datatypes, too -- incoming...

dannon committed:
More styling of the header, include state and collapsible toggle

dannon committed:
Datset component iframe loading indicator fixes

Clearer indication of datset preview

dannon committed:
Adjust headline to match card font-weights -- bold title, normal hid

dannon committed:
Minor animation for dataset header in dataset component

dannon committed:
Refactor header state management to use persistent toggle

dannon committed:
Swap configuration of default dataset viz to be specified inline with the datatype.

dannon committed:
Fix scrolling in dataset component non-iframe tabs -- preview override got dropped somewhere

dannon committed:
swap to standard loading span

dannon committed:
Suggested approach in review for tab management

dannon committed:
Keep enum for tabs

dannon committed:
Fix active tab binding logic.

dannon committed:
Overhaul color application; use theme vars where possible.

dannon committed:
Refactor state styles to use predefined vars.

dannon committed:
Fixed header animation flickering during tab changes by replacing CSS

animations with Vue transitions for smoother state management. This
ensures the header details animation only occurs when intentionally
toggling the header, not during tab navigation.

dannon committed:
Improve dataset header collapse animation to create a smoother, more
immediate transition effect. Replace transform-based animations with
comprehensive transitions for all properties including max-height,
margins, and padding for a fluid collapsing behavior.

dannon committed:
Post-rebase cleanup.

dannon committed:
Don't load top-level visualization_mappings in datatypes_conf -- we pushed these to indivdual datatypes.

dannon committed:
Remove viz default params linkage.  Update datasetview loading state
handling of viz frame.

dannon committed:
Drop 'default_params' concept for preferred viz, for now -- this is extra admin/configuration burden without much benefit at this point

dannon committed:
Add default visualization plugins for GeoJSON and VitessceJSON datatypes; include TIFF viewer for TIFF datatype

guerler commented on lib/galaxy/config/sample/datatypes_conf.xml.sample:
```diff
@@ -142,8 +142,12 @@
-    <datatype extension="geojson" type="galaxy.datatypes.text:GeoJson" mimetype="application/json" display_in_upload="True"/>
-    <datatype extension="vitesscejson" type="galaxy.datatypes.text:VitessceJson" mimetype="application/json" display_in_upload="True"/>
+    <datatype extension="geojson" type="galaxy.datatypes.text:GeoJson" mimetype="application/json" display_in_upload="True">
+      <visualization plugin="openlayers" />
```
This should be Kepler.gl. Is there a reason we don't just add this as `datatype` attribute: `visualization="kepler"`?

guerler replied:
Discussed this in the team meeting and decided to keep it as is, as separate tag.

guerler replied:
Also recommending to keep using `openlayers` as default view for geojson since it loads faster than Kepler. Thx again for all these additions.

dannon committed:
Drop default viz params from DatasetView test, skip frame testing (for
now) -- this is clunky and overly stubbed anyway and we should have an
integration test instead.

dannon committed:
Drop unused import

guerler commented on lib/galaxy/webapps/galaxy/api/datatypes.py:
```diff
@@ -165,3 +236,36 @@ async def edam_data_detailed(self):
+
+    @router.get(
+        "/api/datatypes/{datatype}/visualizations",
+        public=True,
```
If the preferred visualization is provided as a datatype attribute, we would avoid having to make individual requests for each datatype, eliminating the need for additional endpoints. It would also dramatically reduce the complexity of this implementation.

dannon replied:
We could strap it onto the index but there are 500 datatypes on main and once we starting to add extra attributes to that top level request it is going to get messy -- it just shifts complexity from one place to another.  I think this is a reasonable separation, and, as implemented, there are additional details provided at the datatype endpoint that will be useful for showing comprehensive information about specific datatypes, which we want to have.

`/datatypes/` should be just the list of the resource -- it's a simple flat list of available datatypes with no attributes.
```
["4dn_pairs","4dn_pairsam","a3m","ab1","affybatch","agilentbrukeryep.d.tar","agilentmasshunter.d.tar","agp","ampvis2","analyze75","arff","asn1","asn1-binary" .... ]
```
`/datatypes/<datatype>/` is a detail view of the datatype.   I could drop the /visualizations/ sub-filter if that's a problem, but I don't think it adds additional complexity and simply tailors the data provided to a specific use case.
```
{"extension":"h5","description":null,"display_in_upload":true,"mimetype":"application/octet-stream","is_binary":true,"edam_format":"format_3590","edam_data":"data_0006","preferred_visualization":{"visualization":"h5web"}}
```

guerler replied:
Even if we add a detailed `/datatypes/<datatype>` endpoint, I still think the visualization attribute belongs on the top-level /datatypes/ list. Adding a single string to each datatype doesn’t really add complexity, especially compared to creating new endpoints or nested structures. Separating it out only makes sense if we expect multiple visualizations per datatype or need extra metadata—which isn’t the case here.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/datatypes.py:
```diff
@@ -165,3 +236,36 @@ async def edam_data_detailed(self):
         EDAM data contains the EDAM iri, label, and definition."""
         return view_edam_data(self.datatypes_registry, True)
+
+    @router.get(
```
Since there's no consumer I think we can drop this for now.

dannon committed:
Drop all-datatypes visualization mapping endpoint

We have this for edam data, formats, sniffers, etc., though.  It's a
pattern we should make a decision on -- these may be better implemented
as a set of requested keys on the main datatypes summary and exposed
independently on the datatype detail endpoint that has been added.

dannon committed:
rebuild api schema

mvdbeek commented on lib/galaxy/webapps/galaxy/api/datatypes.py:
```diff
@@ -73,6 +79,71 @@ async def index(
         return view_index(self.datatypes_registry, extension_only, upload_only)
 
+    @router.get(
+        "/api/datatypes/{datatype}",
```
I think there's a conflict here with https://github.com/galaxyproject/galaxy/pull/20190/files#diff-7d2152a780aa7c049cf34735fcc435ee1a462dbe4761856524bb87700cedbb17R148 ... move this to the bottom of the file and I think that'll solve the API issues

dannon committed:
DatasetView -- swap order of viz and details

dannon committed:
Relocate rerun to details.

mvdbeek committed:
Selenium fixes

mvdbeek committed:
Fix dataset edit selenium tests

mvdbeek committed:
Keep using history panel item edit button in selenium

New tab needs to be disabled if dataset is running, can follow up later.

mvdbeek committed:
Restore edit button in history content item options

mvdbeek committed:
Wait for ... to disappear

mvdbeek committed:
Move `/api/datatypes/{datatype}` to avoid conflict

dannon committed:
Fix displayUrl type issue

dannon committed:
Fix vue-tsc tabindex issue -- ensure exists

dannon committed:
Fix vue-tsc emit definition issue

dannon committed:
linting fixes in selectdataset and formrecordfielddefinitions (from upstream)

mvdbeek committed:
Selenium test fixes

mvdbeek committed:
Always display eye icon

mvdbeek committed:
Fix test_charts_h5web, drop test_charts_nvd3_bar since there's no nvd3_bar anymore

mvdbeek committed:
Fix tour selectors

mvdbeek committed:
Drop unused buttons, fix visualizations tour
